File: src/lib/base32.test.ts
import { describe, expect, it } from 'vitest';
import { Base32Error, decodeBase32, encodeBase32 } from './base32';

const ascii = (text: string): Uint8Array => new TextEncoder().encode(text);
const toText = (bytes: Uint8Array): string => new TextDecoder().decode(bytes);

/**
 * Die offiziellen Testvektoren aus RFC 4648, Abschnitt 10.
 * Sie decken genau die interessanten Fälle ab: jeden möglichen Rest beim
 * Auffüllen des 40-Bit-Blocks.
 */
const RFC4648_VECTORS: ReadonlyArray<readonly [plain: string, encoded: string]> = [
  ['', ''],
  ['f', 'MY======'],
  ['fo', 'MZXQ===='],
  ['foo', 'MZXW6==='],
  ['foob', 'MZXW6YQ='],
  ['fooba', 'MZXW6YTB'],
  ['foobar', 'MZXW6YTBOI======'],
];

describe('encodeBase32 — RFC 4648 Abschnitt 10', () => {
  for (const [plain, encoded] of RFC4648_VECTORS) {
    it(`codiert "${plain}" zu "${encoded}"`, () => {
      expect(encodeBase32(ascii(plain))).toBe(encoded);
    });
  }

  it('lässt das Padding auf Wunsch weg', () => {
    expect(encodeBase32(ascii('foo'), { padding: false })).toBe('MZXW6');
    expect(encodeBase32(ascii('foobar'), { padding: false })).toBe('MZXW6YTBOI');
  });
});

describe('decodeBase32 — RFC 4648 Abschnitt 10', () => {
  for (const [plain, encoded] of RFC4648_VECTORS) {
    if (encoded === '') continue; // leere Eingabe ist bei uns ein Fehler, siehe unten
    it(`decodiert "${encoded}" zu "${plain}"`, () => {
      expect(toText(decodeBase32(encoded))).toBe(plain);
    });
  }
});

describe('decodeBase32 — Round-trip', () => {
  it('führt für alle Längen 1…40 zurück zum Ausgangs-Byte-Array', () => {
    for (let length = 1; length <= 40; length++) {
      // Deterministische Pseudo-Bytes: reproduzierbar, aber über den ganzen
      // Wertebereich verteilt.
      const original = Uint8Array.from({ length }, (_, i) => (i * 37 + 11) % 256);
      expect(decodeBase32(encodeBase32(original))).toEqual(original);
    }
  });

  it('funktioniert auch ohne Padding', () => {
    for (let length = 1; length <= 20; length++) {
      const original = Uint8Array.from({ length }, (_, i) => (i * 53 + 7) % 256);
      expect(decodeBase32(encodeBase32(original, { padding: false }))).toEqual(original);
    }
  });
});

describe('decodeBase32 — Toleranz gegenüber echten Eingaben', () => {
  const expected = decodeBase32('JBSWY3DPEHPK3PXP');

  it('akzeptiert Kleinbuchstaben', () => {
    expect(decodeBase32('jbswy3dpehpk3pxp')).toEqual(expected);
  });

  it('akzeptiert gemischte Schreibweise', () => {
    expect(decodeBase32('JbSwY3dPeHpK3pXp')).toEqual(expected);
  });

  it('akzeptiert Leerzeichen als Lesehilfe', () => {
    expect(decodeBase32('JBSW Y3DP EHPK 3PXP')).toEqual(expected);
  });

  it('akzeptiert Bindestriche und Zeilenumbrüche', () => {
    expect(decodeBase32('JBSW-Y3DP-EHPK-3PXP')).toEqual(expected);
    expect(decodeBase32('JBSWY3DP\nEHPK3PXP')).toEqual(expected);
    expect(decodeBase32('  JBSWY3DPEHPK3PXP  ')).toEqual(expected);
  });

  it('akzeptiert fehlendes und vorhandenes Padding gleichermassen', () => {
    expect(decodeBase32('MZXW6===')).toEqual(decodeBase32('MZXW6'));
    expect(decodeBase32('MZXW6YQ=')).toEqual(decodeBase32('MZXW6YQ'));
  });

  it('decodiert das RFC-Test-Secret zurück nach ASCII', () => {
    // Dieses Secret taucht in RFC 4226 und RFC 6238 auf — und in der Demo der App.
    expect(toText(decodeBase32('GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ'))).toBe(
      '12345678901234567890',
    );
  });
});

describe('decodeBase32 — Fehlerfälle', () => {
  it('lehnt eine leere Eingabe ab', () => {
    expect(() => decodeBase32('')).toThrow(Base32Error);
    expect(() => decodeBase32('   ')).toThrow(/leer/i);
    expect(() => decodeBase32('====')).toThrow(/leer/i);
  });

  it('lehnt Zeichen ausserhalb des Alphabets ab', () => {
    // 0, 1 und 8 gibt es in Base32 nicht.
    expect(() => decodeBase32('ABCD0FGH')).toThrow(Base32Error);
    expect(() => decodeBase32('ABCD1FGH')).toThrow(Base32Error);
    expect(() => decodeBase32('ABCD8FGH')).toThrow(Base32Error);
    expect(() => decodeBase32('ABCD$FGH')).toThrow(Base32Error);
    expect(() => decodeBase32('ÄBCDEFGH')).toThrow(Base32Error);
  });

  it('nennt das störende Zeichen in der Fehlermeldung', () => {
    expect(() => decodeBase32('JBSW0Y3D')).toThrow(/»0«/);
    expect(() => decodeBase32('JBSW0Y3D')).toThrow(/Stelle 5/);
  });

  it('lehnt Padding mitten im String ab', () => {
    expect(() => decodeBase32('MZXW6===YTB')).toThrow(/=/);
  });

  it('lehnt unmögliche Längen ab', () => {
    // Reste 1, 3 und 6 (mod 8) können nie entstanden sein.
    expect(() => decodeBase32('A')).toThrow(/Länge/);
    expect(() => decodeBase32('ABC')).toThrow(/Länge/);
    expect(() => decodeBase32('ABCDEF')).toThrow(/Länge/);
    expect(() => decodeBase32('ABCDEFGHA')).toThrow(/Länge/);
  });

  it('akzeptiert alle möglichen Längen', () => {
    for (const validLength of [2, 4, 5, 7, 8, 10, 16, 26, 32]) {
      const input = 'A'.repeat(validLength);
      expect(() => decodeBase32(input)).not.toThrow();
      expect(decodeBase32(input)).toHaveLength(Math.floor((validLength * 5) / 8));
    }
  });
});

File: android-native/app/src/test/kotlin/io/github/keco216/clockwork/core/Base32Test.kt
package io.github.keco216.clockwork.core

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Portiert aus `src/lib/base32.test.ts` — Fall fuer Fall, dieselben Vektoren.
 */
class Base32Test {

    private fun ascii(text: String) = text.toByteArray(Charsets.US_ASCII)
    private fun text(bytes: ByteArray) = String(bytes, Charsets.UTF_8)

    /**
     * Die offiziellen Testvektoren aus RFC 4648, Abschnitt 10. Sie decken
     * genau die interessanten Faelle ab: jeden moeglichen Rest beim Auffuellen
     * des 40-Bit-Blocks.
     */
    private val rfc4648 = listOf(
        "" to "",
        "f" to "MY======",
        "fo" to "MZXQ====",
        "foo" to "MZXW6===",
        "foob" to "MZXW6YQ=",
        "fooba" to "MZXW6YTB",
        "foobar" to "MZXW6YTBOI======",
    )

    @Test
    fun `die Tabelle enthaelt alle 7 Vektoren aus Abschnitt 10`() {
        // Wie bei HOTP: Eine Schleife ueber eine zu kurze Tabelle laeuft gruen
        // durch. Diese Zeile gibt den beiden Vektor-Tests ihre Aussage.
        assertEquals(7, rfc4648.size)
    }

    @Test
    fun `encodeBase32 erfuellt RFC 4648 Abschnitt 10`() {
        var checked = 0
        for ((plain, encoded) in rfc4648) {
            assertEquals("\"$plain\"", encoded, encodeBase32(ascii(plain)))
            checked++
        }
        assertEquals(7, checked)
    }

    @Test
    fun `encodeBase32 laesst das Padding auf Wunsch weg`() {
        assertEquals("MZXW6", encodeBase32(ascii("foo"), padding = false))
        assertEquals("MZXW6YTBOI", encodeBase32(ascii("foobar"), padding = false))
    }

    @Test
    fun `decodeBase32 erfuellt RFC 4648 Abschnitt 10`() {
        for ((plain, encoded) in rfc4648) {
            // Die leere Eingabe ist bei uns ein Fehler, siehe unten.
            if (encoded.isEmpty()) continue
            assertEquals("\"$encoded\"", plain, text(decodeBase32(encoded)))
        }
    }

    @Test
    fun `Round-trip fuer alle Laengen 1 bis 40`() {
        for (length in 1..40) {
            // Deterministische Pseudo-Bytes: reproduzierbar, aber ueber den
            // ganzen Wertebereich verteilt.
            val original = ByteArray(length) { ((it * 37 + 11) % 256).toByte() }
            assertArrayEquals("Laenge $length", original, decodeBase32(encodeBase32(original)))
        }
    }

    @Test
    fun `Round-trip funktioniert auch ohne Padding`() {
        for (length in 1..20) {
            val original = ByteArray(length) { ((it * 53 + 7) % 256).toByte() }
            assertArrayEquals(
                "Laenge $length",
                original,
                decodeBase32(encodeBase32(original, padding = false)),
            )
        }
    }

    @Test
    fun `akzeptiert Kleinbuchstaben und gemischte Schreibweise`() {
        val expected = decodeBase32("JBSWY3DPEHPK3PXP")
        assertArrayEquals(expected, decodeBase32("jbswy3dpehpk3pxp"))
        assertArrayEquals(expected, decodeBase32("JbSwY3dPeHpK3pXp"))
    }

    @Test
    fun `akzeptiert Leerzeichen Bindestriche und Zeilenumbrueche`() {
        val expected = decodeBase32("JBSWY3DPEHPK3PXP")
        assertArrayEquals(expected, decodeBase32("JBSW Y3DP EHPK 3PXP"))
        assertArrayEquals(expected, decodeBase32("JBSW-Y3DP-EHPK-3PXP"))
        assertArrayEquals(expected, decodeBase32("JBSWY3DP\nEHPK3PXP"))
        assertArrayEquals(expected, decodeBase32("  JBSWY3DPEHPK3PXP  "))
    }

    @Test
    fun `akzeptiert fehlendes und vorhandenes Padding gleichermassen`() {
        assertArrayEquals(decodeBase32("MZXW6"), decodeBase32("MZXW6==="))
        assertArrayEquals(decodeBase32("MZXW6YQ"), decodeBase32("MZXW6YQ="))
    }

    @Test
    fun `decodiert das RFC-Test-Secret zurueck nach ASCII`() {
        // Dieses Secret taucht in RFC 4226 und RFC 6238 auf — und im
        // Testschluessel-Knopf der App.
        assertEquals(
            "12345678901234567890",
            text(decodeBase32("GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ")),
        )
    }

    @Test
    fun `lehnt eine leere Eingabe ab`() {
        assertKey("err.base32.empty") { decodeBase32("") }
        assertKey("err.base32.empty") { decodeBase32("   ") }
        assertKey("err.base32.empty") { decodeBase32("====") }
    }

    @Test
    fun `lehnt Zeichen ausserhalb des Alphabets ab`() {
        // 0, 1 und 8 gibt es in Base32 nicht.
        for (input in listOf("ABCD0FGH", "ABCD1FGH", "ABCD8FGH", "ABCD\$FGH", "ÄBCDEFGH")) {
            assertKey("err.base32.badChar", input) { decodeBase32(input) }
        }
    }

    @Test
    fun `nennt das stoerende Zeichen und seine Stelle`() {
        val error = capture { decodeBase32("JBSW0Y3D") }
        assertEquals("err.base32.badChar", error.key)
        assertEquals("0", error.args["char"])
        assertEquals("5", error.args["position"])
    }

    @Test
    fun `lehnt Padding mitten im String ab`() {
        assertKey("err.base32.paddingInside") { decodeBase32("MZXW6===YTB") }
    }

    @Test
    fun `lehnt unmoegliche Laengen ab`() {
        // Reste 1, 3 und 6 (mod 8) koennen nie entstanden sein.
        for (input in listOf("A", "ABC", "ABCDEF", "ABCDEFGHA")) {
            assertKey("err.base32.badLength", input) { decodeBase32(input) }
        }
    }

    @Test
    fun `bleibt bei pathologischen Eingaben linear schnell`() {
        // Regressionstest aus dem Web: Dort hatte `cleaned.replace(/=+$/, '')`
        // quadratische Laufzeit (80 000 Zeichen ~ 1,8 s — der Tab friert ein).
        // Javas Regex-Maschine hat dasselbe Rueckzugsverhalten, deshalb laeuft
        // das Abschneiden auch hier ueber eine Schleife.
        val start = System.nanoTime()
        for (length in listOf(20_000, 200_000)) {
            // Das Padding steht hier VORNE, das letzte Zeichen ist ein 'A' —
            // abgeschnitten wird also nichts, und der Fehler ist folgerichtig
            // "= mitten drin" und nicht "ungueltiges Zeichen".
            assertKey("err.base32.paddingInside") { decodeBase32("=".repeat(length) + "A") }
        }
        val millis = (System.nanoTime() - start) / 1_000_000
        // Grosszuegige Schranke: linear braucht das wenige Millisekunden,
        // quadratisch waeren es Minuten.
        assertTrue("Laufzeit $millis ms", millis < 1000)
    }

    @Test
    fun `akzeptiert alle moeglichen Laengen`() {
        for (validLength in listOf(2, 4, 5, 7, 8, 10, 16, 26, 32)) {
            val input = "A".repeat(validLength)
            assertEquals("Laenge $validLength", validLength * 5 / 8, decodeBase32(input).size)
        }
    }
}
